fuzzing/tools/dict_validator.py

```diff
@@ -0,0 +1,89 @@
+#
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""
+This script is used to validate the format of the fuzzing dictionary
+"""
+
+from absl import app
+from absl import flags
+from string import hexdigits
+
+FLAGS = flags.FLAGS
+
+
+# Validate a single entry in the dictionary
+def validate_entry(line):
+    line = line.strip()
+    if not line or line.startswith('#'):
+        return True
+    if len(line) < 2 or line[-1] != '"':
+        return False
+
+    left, right = 0, len(line) - 1
+    # Find the opening "
+    while left < right and line[left] != '"':
+        left += 1
+
+    if left >= right:
+        return False
+
+    # Use set to contain hex digits to decrease the query time complexity
+    hex_set = set(hexdigits)
+    while left < right:
+        left += 1
+        chr = line[left]  # Single character
+
+        if not (chr.isprintable() or chr.isspace()):
+            return False
+
+        # Handle '\\'
+        if chr == '\\':
+            if left + 1 <= right and (line[left + 1] == '\\' or
+                                      line[left + 1] == '"'):
+                left += 1
+                continue
+
+            # Handle '\xAB'
```

Comment:
I am wondering if we have any existed tools validating the dictionary entry? 

The codes here are modeling  [libfuzzer dictionary parser](https://github.com/llvm/llvm-project/blob/6c67ee0f58323fe39db84635ef8cea629c2214b4/compiler-rt/lib/fuzzer/FuzzerUtil.cpp#L81). It allows `"ab""` to pass the check while filters out `"\xA"`, those look strange to me.

Reply:
Yep, libfuzzer's own dictionary parser is probably the gold standard. You can test out some of these cases by running any fuzzer with `-dict`, and it will tell you if there are any dictionary errors, and how many successful dictionary entries there were. I tried the case above, and found
```-----------------------------------------------------------------------------
ParseDictionaryFile: error in line 2
		"\xA"
```

It is definitely worth adding these testcases to the test dictionaries, and running a fuzzer with the test dictionaries to make sure the valid ones are parsed, and each invalid one isn't

Reply:
Yes, I have created `valid.dict` and `invalid.dict` for the valid and invalid dictionary entries respectively under `fuzzing/tools/dict_data`.

For the `"ab""` with which I was confused, it turns out to be a valid entry.

Reply:
Ah yeah, sorry -- I meant adding "\xA" to the `valid.dict` and also just verifying that all of valid.dict gets accepted to libfuzzer by running the fuzzer with it!